Packs/Phishing/Scripts/PhishingDedupPreprocessingRule/PhishingDedupPreprocessingRule.py

```diff
@@ -0,0 +1,209 @@
+        existing_incidents_df = existing_incidents_df[mask]
+    existing_incidents_df.sort_values(by='similarity', ascending=False, inplace=True)
+    if len(existing_incidents_df) > 0:
+        return existing_incidents_df.iloc[0], existing_incidents_df.iloc[0]['similarity']
+    else:
+        return None, None
+
+
+def close_new_incident_and_link_to_existing(new_incident, existing_incident, similarity):
+    entries = []
+    # dumpable_new_incident = {k: v for k, v in new_incident.items() if not isinstance(v, (np.floating, np.integer))}
+    entries.append({'Contents': "Duplicate incident: " + new_incident['name']})
+    entries.append({"Type": entryTypes['note'],
+                    "ContentsFormat": "json",
+                    "Contents": json.dumps(new_incident)
+                    })
+    entries_str = json.dumps(entries)
+    demisto.executeCommand("addEntries", {"id": existing_incident["id"], "entries": entries_str})
+    res = demisto.executeCommand("linkIncidents", {
+        'linkedIncidentIDs': new_incident['id'],
+        'incidentId': existing_incident['id']})
+    if is_error(res):
+        return_error(res)
+    demisto.results('Duplicated incident found: {}, {}'.format(new_incident['id'], existing_incident['id']))
+
+
+def create_new_incident():
+    demisto.results('No duplicated incident found')
+
+
+def main():
+    global EMAIL_BODY_FIELD, EMAIL_SUBJECT_FIELD, EMAIL_HTML_FIELD, FROM_FIELD, MIN_TEXT_LENGTH, FROM_POLICY
+    input_args = demisto.args()
+    EMAIL_BODY_FIELD = input_args.get('emailBody', EMAIL_BODY_FIELD)
+    EMAIL_SUBJECT_FIELD = input_args.get('emailSubject', EMAIL_SUBJECT_FIELD)
+    EMAIL_HTML_FIELD = input_args.get('emailBodyHTML', EMAIL_HTML_FIELD)
+    FROM_FIELD = input_args.get('emailFrom', FROM_FIELD)
+    FROM_POLICY = input_args.get('fromPolicy', FROM_POLICY)
+    existing_incidents = get_existing_incidents(input_args)
+    if len(existing_incidents) == 0:
+        create_new_incident()
+        return
+    new_incident = demisto.incidents()[0]
+    new_incident_df = preprocess_incidents_df([new_incident])
+    existing_incidents_df = preprocess_incidents_df(existing_incidents)
+    existing_incidents_df = filter_out_new_incident(existing_incidents_df, new_incident)
+
+    if len(existing_incidents_df) == 0 or len(
+            new_incident_df) == 0:  # len(new_incident_df)==0 means new incident is too short
+        create_new_incident()
+        return
+    new_incident_preprocessed = new_incident_df.iloc[0].to_dict()
+    duplicate_incident_row, similarity = find_duplicate_incidents(new_incident_preprocessed,
+                                                                  existing_incidents_df)
+    if duplicate_incident_row is None or similarity < SIMILARITY_THRESHOLD:
+        create_new_incident()
+    else:
+        return close_new_incident_and_link_to_existing(new_incident, duplicate_incident_row, similarity)
+
```

Comment:
should we return True?

Reply:
If it's a pre-processing script - yes
Since currently it can't be a pre-processing script, I preferred it to be a working regular script

Reply:
sure